File: src/app/shared/service/api/product.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams, HttpRequest } from '@angular/common/http';
import { ProductDTORequest } from '../../model/request/productDTORequest';
import { Observable } from 'rxjs';
import { environment } from 'src/environments/environment';

@Injectable({
    providedIn: 'root'
})

export class ProductService {
    constructor(
        private _httpClient: HttpClient
    ) { }

    findByKeyWord(key_word: string, page: number, size: number): Observable<any> {
        let queryParameters = new HttpParams();
        queryParameters = queryParameters.append('page', page);
        queryParameters = queryParameters.append('size', size);
        queryParameters = queryParameters.append('field', '');
        queryParameters = queryParameters.append('order', '');
        queryParameters = queryParameters.append('key_word', key_word);
        return this._httpClient.get<any>(environment.urlBase + '/products/product?' + queryParameters);
    }

    public findById(id: number): Observable<any> {
        return this._httpClient.get<any>(environment.urlBase + '/products/product/' + id);
    }

    public save(product: ProductDTORequest): Observable<any> {
        return this._httpClient.post<any>(environment.urlBase + '/products/product', product);
    }

    public update(product: ProductDTORequest, id: number): Observable<any> {
        return this._httpClient.put<any>(environment.urlBase + '/products/product/' + id, product);
    }

    public delete(id: number): Observable<any> {
        return this._httpClient.patch<any>(environment.urlBase + '/products/product/' + id, null);
    }

    public filterAutoComplete(key_word:string): Observable<any>{
        return this._httpClient.get(environment.urlBase + '/products/product/filter?key_word='+key_word);
    }
}

File: src/main/java/com/sys/botica/crce/pe/sys_botica/service/ProductService.java
package com.sys.botica.crce.pe.sys_botica.service;

import java.util.List;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import com.sys.botica.crce.pe.sys_botica.dto.HrefEntityDTO;
import com.sys.botica.crce.pe.sys_botica.dto.ProductDTO;
import com.sys.botica.crce.pe.sys_botica.dto.request.ProductDTORequest;

public interface ProductService {
	public HrefEntityDTO save(ProductDTORequest product);
	public HrefEntityDTO update(ProductDTORequest product,Long id);
	public HrefEntityDTO delete(Long id);
	public Page<ProductDTO> findByKeyWord(String key_word,Pageable pageable);
	public ProductDTO findByID(Long id);
	public List<ProductDTO> findByKeyWordSQL(String key_word);
}
